src/misc/cross.cpp

```diff
@@ -715,6 +715,88 @@ std::string cfstr_to_string(CFStringRef source)
 	return lang;
 }
 
+#if defined (__linux__)
+
+#include <fcntl.h>
+
+int get_num_physical_cpus()
+{
+	constexpr int MaxCpus = 128;
+	bool core_ids[MaxCpus] = {};
+	int num_cores = 0;
+	for (int i = 0; i < MaxCpus; ++i) {
+		std::string path = "/sys/devices/system/cpu/cpu" + std::to_string(i) + "/topology/core_id";
+		int fd = open(path.c_str(), O_RDONLY);
+		if (fd != -1) {
+			char core_id_str[6] = {};
+			if (read(fd, core_id_str, 5) > 0) {
+				int core_id = atoi(core_id_str);
+				if (core_id >= 0 && core_id < MaxCpus && !core_ids[core_id]) {
+					core_ids[core_id] = true;
+					++num_cores;
+				}
+			}
+			close(fd);
+		}
+	}
+	return std::max(num_cores, 1);
+}
+
+#elif defined (MACOSX)
+
+#include <sys/sysctl.h>
+
+int get_num_physical_cpus()
+{
+	int num_cpus = 0;
+	size_t size = sizeof(num_cpus);
+	sysctlbyname("hw.physicalcpu", &num_cpus, &size, nullptr, 0);
+	return std::max(num_cpus, 1);
+}
```

Comment:
This Mac path is untested by the way.  Apple wins with this API being by far the simplest so hopefully it just works.

Reply:
@weirddan455 
With this code I'm getting the following on my M1 MacBook:

![image](https://github.com/user-attachments/assets/fca84670-4e6b-4b87-9116-26edac23b186)

So it reports 10, whereas it should only report 8 I think.

I can get the number of performance cores (8) by querying `hw.perflevel0.physicalcpu`. See here:
https://forums.developer.apple.com/forums/thread/692671

<img width="1218" alt="image" src="https://github.com/user-attachments/assets/c0df67d8-20ca-4269-8397-dd7c1af87fcc">

Reply:
Oh I didn't think about E-cores.  They're still physical cores so I assume the API is working as expected.  Unsure if we want to be spinning up extra threads for E cores or not.  I'm pretty sure Windows and Linux count E cores as physical cores on the newer Intel CPUs as well so we'll have to make the same decision for x86.  I don't have one of those CPUs to benchmark on though.

I guess the M1 chips don't have a hyper-threading equivalent?  That was what this was trying to remove from the count.

Reply:
> I guess the M1 chips don't have a hyper-threading equivalent? That was what this was trying to remove from the count.

Apparently they don't:
https://www.quora.com/Why-doesnt-Apple-use-Hyper-Threaded-CPUs-instead-of-multiple-cores-like-other-companies-do

Cool, then your original code is working as expected 👍🏻

Reply:
My hunch says we should probably exclude E cores but without hardware I'm just guessing really.

It won't actually make a difference on your laptop or the high end Intel CPUs since we're clamping to a maximum of 8 threads.  It would mostly be these Core i5s that have 6 P-cores and 4 or 8 E-cores:  https://en.wikipedia.org/wiki/Raptor_Lake#List_of_14th_generation_Raptor_Lake_processors

I'm also not sure if the OS prefers to schedule these high-usage threads on E-cores or the logical (hyperthreaded sibling) P-cores.

Reply:
Yeah, I'm fine to leave it as-is for now, then complicate it later once we have data 😆

Reply:
I'll leave this as is for now.  I'm not seeing an easy way to identify E cores without comparing frequencies or cache information.  I don't feel confident in such code working without me being able to test on real hardware.  This can always be adjusted later.